demos/http/http_demo_s3_download_async/http_demo_s3_download_async.c

```diff
@@ -0,0 +1,751 @@
+    struct mq_attr queueSettings;
+
+    /* Number of requests made */
+    size_t requestCount = 0;
+
+    /* PID of HTTP thread. */
+    pid_t httpThread = -1;
+
+    ( void ) argc;
+    ( void ) argv;
+
+    for( ; ; )
+    {
+        LogInfo( ( "HTTPS Client Asynchronous S3 download demo using pre-signed URL:\n%s", S3_PRESIGNED_URL ) );
+
+        /**************************** Parse Signed URL. ******************************/
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            /* Retrieve the path location from S3_PRESIGNED_URL. This function returns the length of the path
+             * without the query into pathLen. */
+            httpStatus = getUrlPath( S3_PRESIGNED_URL,
+                                     strlen( S3_PRESIGNED_URL ),
+                                     &pPath,
+                                     &pathLen );
+
+            if( httpStatus != HTTP_SUCCESS )
+            {
+                LogError( ( "An error occurred in getUrlPath() on URL %s. Error code: %d",
+                            S3_PRESIGNED_URL,
+                            httpStatus ) );
+                returnStatus = EXIT_FAILURE;
+            }
+        }
+
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            /* Retrieve the address location and length from the S3_PRESIGNED_URL. */
+            httpStatus = getUrlAddress( S3_PRESIGNED_URL,
+                                        strlen( S3_PRESIGNED_URL ),
+                                        &pAddress,
+                                        &serverHostLength );
+
+            if( httpStatus != HTTP_SUCCESS )
+            {
+                LogError( ( "An error occurred in getUrlAddress() on URL %s\r\n. Error code %d",
+                            S3_PRESIGNED_URL,
+                            httpStatus ) );
+                returnStatus = EXIT_FAILURE;
+            }
+        }
+
+        /**************************** Connect. ******************************/
+
+        /* Establish TLS connection on top of TCP connection using OpenSSL. */
+        if( returnStatus == EXIT_SUCCESS )
+        {
+            /* Attempt to connect to the HTTP server. If connection fails, retry after
+             * a timeout. Timeout value will be exponentially increased till the maximum
+             * attempts are reached or maximum timeout value is reached. The function
+             * returns EXIT_FAILURE if the TCP connection cannot be established to
+             * broker after configured number of attempts. */
+            returnStatus = connectToServerWithBackoffRetries( connectToServer,
+                                                              &networkContext );
+
+            if( returnStatus == EXIT_FAILURE )
+            {
+                /* Log error to indicate connection failure after all
+                 * reconnect attempts are over. */
+                LogError( ( "Failed to connect to HTTP server %s.",
+                            serverHost ) );
```

Comment:
As much as possible, the function doing the work should be responsible for the logging.